### PR TITLE
Search API redo

### PR DESCRIPTION
**Pull request by SabrinaJaye**
This is for satisfying #141

- [x] An exact character-like search. `findfirst(DNA_A, dna"GATC")`, which returns nothing or an integer.
     - [x] `findnext`
     - [x] `findfirst`
     - [x] `findlast`
     - [x] `findprev`
- [x] An exact string-like search. `findfirst(MatchSearchQuery(dna"a"), dna"GATC")`, which returns nothing or a range.
     - [x] `findnext`
     - [x] `findfirst`
     - [x] `findlast`
     - [x] `findprev`
- [x] A search that matches/considers ambiguous symbols. `findfirst(CompatibleSearchQuery(pat), seq)`, which returns nothing or a range.
     - [x] `findnext`
     - [x] `findfirst`
     - [x] `findlast`
     - [x] `findprev`
- [x] A search that matches/considers ambiguous symbols and allows n mismatches. `findfirst(ApproximateSearchQuery(pat, n; rev=false), seq)`, which returns nothing or a range.
     - [x] `findnext`
     - [x] `findfirst`
     - [x] `findlast`
     - [x] `findprev`
- [x] A search that matches a PWM to a given threshold. `findfirst(PWMSearchQuery(pwm, threshold), seq)`, which returns nothing or a range.
     - [x] `findnext`
     - [x] `findfirst`
     - [x] `findlast`
     - [x] `findprev`
## :ballot_box_with_check: Checklist

- [x] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/).
- [x] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/v1/manual/documentation/).
- [x] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [x] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [x] :ok: All changes should be compatible with the latest stable version of Julia.
- [x] :thought_balloon: I have commented liberally for any complex pieces of internal code.

### CODE REVIEW

**Comment by CiaranOMara** [comment]
Looking over [`Base.findfirst`](https://github.com/JuliaLang/julia/blob/ae8452a9e0b973991c30f27beb2201db1b0ea0d3/base/strings/search.jl#L104-L105), it generally appears to be a special case of `findnext`.  Perhaps our `findfirst` methods could initially be renamed as `findnext`, then adjusted for #141?

**Comment by CiaranOMara** [comment]
Seems like an optimisation issue. I'm assuming solution 1 is easier to implement, but not the most elegant. I think it would be nice to get to solution 2. Though it might not be useful to go at it directly. Maybe it'd be better to implement solution 1 and make sure the high-level interface feels nice, then optimise towards solution 2, which could be a patch or feature release.

**Comment by jakobnissen** [comment]
Hmm, looking at the approximate search source code, there are a few more issues:
* The `fPcom` and `bPcom` vectors are abstractly typed, because they must hold an integer type with at least as many bits as the input query. This also implies that searching for sequences longer than 128 symbols is very slow because it reverts to `BigInt`. My instinct would be to cap the search query at 64 symbols and always use en element type of `UInt64`. In my experience, if I want to search for sequences longer than 30-ish bp, I would use pairwise alignment algorithms, so that will not be an issue. But maybe that's too limiting?
* If the above mentioned integer size is not BigInt but a BitInteger type, `approx_preprocess` can be effectively calculated in one pass using the code below
```
function approx_preprocess(pat)
	m = length(pat)
	m > 128 && error("ApproxSearchQuery can only be used for sequences at most 64 symbols in length")
	fw = zeros(UInt64, length)
	Σ = alphabet(eltype(pat))
	for i in 1:m
		y = pat[i]
		for x in Σ
			if BioSequences.iscompatible(x, y)
				fw[encoded_data(x) + 0x01] |= UInt(1) << (i - 1)
			end
		end
	end
	shift = 64 - m
	bw = [reversebits(i) >>> (shift & 63) for i in fw]
	return fw, bw
end
```
The code assumes `BioSequences.iscompatible` and `encoded_data` has been defined for the relevant `BioSymbol`, that the biosymbol is encoded to an integer type at most 64 bits in size, and that the query sequence is at most 64 symbols. I think we should settle on an interface for BioSymbols at some point to make sure this is satisfied.
If this is implemented, we might as well calculate both passes in one go.

* There's quite a lot of use of `reinterpret(UInt8, x)` in the file. This should be replaced with `encoded_data(x)`, since a user could make a BioSymbol type which is not represented by a single byte.

**Comment by BenJWard** [comment]
The last commit I just had the constructor do both preprocessing passes. On a similar logic to @jakobnissen's comment - how big are these patterns going to be really? Most people will likely be using this to create a query, and then search each of their reads/contigs/genes... The one preprocessing step is not gonna be the bottleneck really!

I've also just had the constructor make a copy of the sequence for the query. Because I could imagine a scenario in which someone does a mutating operation on the very same sequence object they used to make a query, and as a result screwing the query up!

File: src/search/approx.jl [hunk 1]
```diff
@@ -41,24 +40,44 @@ function ApproximateSearchQuery(pat::BioSequence, k::Integer, direction::Symbol
...
         for x in Σ
             if BioSequences.iscompatible(x, y)
-                Pcom[reinterpret(UInt8, x) + 0x01] |= one(T) << (i - 1)
+                fw[encoded_data(x) + 0x01] |= UInt(1) << (i - 1)
             end
         end
     end
-    return Pcom
+    shift = 64 - m
+    bw = [bitreverse(i) >>> (shift & 63) for i in fw]
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/search/approx.jl:59]: If so, we might want to change the CI to include 1.6 as the LTS. If not, we could add a `reversebits` method in `bit-manipulation`.

[the remark this answers is not in the document]
Comment (BenJWard) [reply src/search/approx.jl:59]: Yeah we can inc the required version of julia.

File: src/search/approx.jl [hunk 2]
```diff
@@ -41,24 +40,44 @@ function ApproximateSearchQuery(pat::BioSequence, k::Integer, direction::Symbol
...
 
-function approx_preprocess(pat, forward)
-    # select a bit vector type
-    # TODO: BigInt is very slow, consider implementing "4.2 THE BLOCKS MODEL"
+function approx_preprocess(pat)
     m = length(pat)
-    T = m ≤ 64 ? UInt64 : m ≤ 128 ? UInt128 : BigInt
+    if m ≤ 64
+        throw(ArgumentEror("query pattern sequence must have length of 64 or less"))
+    end
```
[the remark this answers is not in the document]
Comment (BenJWard) [reply src/search/approx.jl]: Yep, another case of knowing what I meant to type, and yet...

**Comment by BenJWard** [comment]
Ok, so approximate searching with k mismatches is implemented, and documented. I've added a note to the manual under the relevant heading to say hey, if you want to fuzzy search long motifs in this way, you're probably looking for BLAST or some other local alignment algorithm.

**Comment by BenJWard** [comment]
@jakobnissen Ok. So the approx search does something I've not noticed before, but now I have, I can't un-see it and it bugs me.

See these tests:

```julia
seq = dna"ACGTACG"

        @test findnext(ApproximateSearchQuery(dna"ACG", 0), seq, 1) === 1:3
        @test findnext(ApproximateSearchQuery(dna"ACG", 1), seq, 1) === 1:2
        @test findnext(ApproximateSearchQuery(dna"ACG", 2), seq, 1) === 1:1
        @test findnext(ApproximateSearchQuery(dna"ACG", 3), seq, 1) === nothing #1:0
        @test findnext(ApproximateSearchQuery(dna"ACG", 4), seq, 1) === nothing #1:0
```

See it's annoying to me the range returned after the first test shrinks, e.g. test 2 is 1:2, not 1:3. Even though, 1:3 is the subsequence that satisfies the query! Same for the 3rd test. Like... ACG is the first sequence, so why is allowing up to 2 mismatches affecting the range I get out! Then the final two tests it gets so silly as to trigger the thing to return `nothing`, even though, a perfect ACG match is literally the first subsequence.

I'm not going mad, it _is_ weird, right?

**Comment by BenJWard** [comment]
Ok, so a draft version of the exact matching algorithm has been added, which is just a modified version of the one that considers compatibility to be a match tbh. I think it could be nicer and code duplication reduced. Either byy making both types of query MatchSearchQuery and CompatibleSearchQuery some subtype of an abstract query, or by folding them both into one query type `ExactQuery{<:BioSequence,F<:Union{typeof(isequal), typeof(iscompatible)}}`, where the query type has a parameter to dispatch on the comparitor function used for preprocessing and searching so you could do something like: `findfirst(exactquery(dna"ACGT", isequal), seq)` or `findfirst(exactquery(dna"ACGT", iscompatible), seq)`.

@jakobnissen I could do with your opinion on this. So the `cbits` variable in both types, is basically used as a tiny bloom filter to check if an element is in the query pattern or not. In the query type that considers compatibility, `cbits` is formed using `compatibits`, because it needs to check if there are compatible symbols in the query pattern. But for the exact matching type, I think actually we can just construct and use the `cbits` bloom filter in the same way that the implementation for `String` in base does: i.e. 

```
function _search_bloom_mask(c)
    UInt64(1) << (c & 63)
end
``` 

Where `c` is a byte from a `String`.

**Comment by CiaranOMara** [comment]
> I'm not going mad, it _is_ weird, right?

I too would expect the resulting range to be the same length as the search pattern or a result of `nothing`.

>  I think it could be nicer and code duplication reduced. Either by making both types of query MatchSearchQuery and CompatibleSearchQuery some subtype of an abstract query

Subtyping the queries from an `AbstractQuery` type will also catch a lot of low hanging fruit. For example, the `findfirst`, `findlast`, and `findall` methods.

```julia
Base.findlast(query::AbstractQuery, seq::BioSequence) = findprev(query, seq, lastindex(seq))
Base.findfirst(query::AbstractQuery, seq::BioSequence) = findnext(query, seq, firstindex(seq))
Base.findall( . . . ) = . . .
```

There are probably others too that don't immediately come to mind.

**Comment by CiaranOMara** [comment]
I reckon the `findnext` and `findprev` methods could become generic if they interact with an `AbstractQuery` type implementing the iterator interface. I haven't thought about this too deeply and am unsure whether such an approach would put us in awkward situations. Perhaps such an approach would lay out some clearly defined expectations of a search query.

**Comment by BenJWard** [comment]
I think for now, I'm just gonna get it to the point it does things in a somewhat sensible manner, so long as the API is there, and then we can deal with internals later, since thoughts on how to generalise things lead to things that may end up being BioSymbol interface issues.

**Comment by jakobnissen** [comment]
@BenJWard Okay a few comments

First, I think this is great. It's much better than the old API, and we should use this. I have a few comments - most of them do not pertain to the API but are internal, so I can add that later as PRs if I can figure out how to. In no particular order

* Any reason `k` must be specified when creating an `ApproxSearchQuery`? The problem with that is that I can easily imagine a user wanting to search with the same query with first 0, then 1, then 2 etc. mismatches until they find a hit. With this API, they need to create N number of queries. If the motivation is to keep the find API more similar to Base (which I support!), then perhaps add a function to create a query from another query with a new K? This can re-use all the internal data except K and so be almost instant. The alternative is to specify `k` in the find* functions.
* You are right there are some code duplication. I think it should be possible to collapse all 4 `quicksearch` methods into one. But as you say, that is not critical and can be done later.
* The `quicksearch` algorithm used for exact match seem to work best when the alphabet is larger, because it allows it to skip more symbols ahead when a match is rejected. That is unfortunate, since RNA and DNA alphabets are small. For example, I just benchmarked finding a short sequence after 70,000 symbols, where it is 7.2 times faster to search for amino acids than DNA. In fact, for exact DNA/RNA queries, I think it's faster to pack the first N symbols into an integer (just like a kmer, just simpler) and check that before continuing the search. I can look into that later, the speed is not too important now!
* It would be nice to have an approximate search query that did not use compat bits but required an exact match. I think we should be able to do it just like you said, by generalizing over `isequal` vs `iscompatible` without too much hassle.
* A small nitpick: Right now, the `ExactSearchQuery` includes the compatible query (which makes it not necessarily exact), whereas the `MarchSearchQuery` actually needs it to be exact. Doesn't it make more sense to switch the names of those two, such that the exact query means exact, and match is ambiguous to mean either exact match or compatible match?

**Comment by jakobnissen** [comment]
@BenJWard okay so I have some time to review it today and tomorrow.
Regarding the ApproxSearch results (i.e. the ones here https://github.com/BioJulia/BioSequences.jl/pull/190#issuecomment-956235944), given the examples you gave, I actually think they are all reasonable, except the ones that return `nothing`.

If you truly want the first subsequence of e.g. `AGCT` which matches `AGC` with up to 2 mismatches, the first subsequence that matches is indeed just the leading `A`. It's counter-intuitive, but I think it's correct.

However, if you ask for the first subsequence that matches `AGC` with 3 mismatches, the correct result should be the empty sequence at `1:0`! Similar to how
```julia
julia> findfirst("", "hello")
1:0
```

What happened here was probably that I got a little too overzealous when converting the empty results to `nothing`, assuming that no empty result was valid. Clearly that was a bad assumption!

I'll look into the other issue now :)

**Comment by CiaranOMara** [comment]
> If you truly want the first subsequence of e.g. AGCT which matches AGC with up to 2 mismatches, the first subsequence that matches is indeed just the leading A.

I think this is a bit too clever as it allows for 2 positional mismatches.

If padding is added to the sequence `ACGTACG`, then a query `xxA` allowing 2 symbol mismatches could return a match containing the first `A`. The strict range for which that match occurs would be `-1:1`. But, the range for which the match occurs is just `1:1`.

A naive, and albeit incomplete implementation of this logic is below.
```julia
using OffsetArrays


seq = "ACGTACG" |> collect
pat = "ACG" |> collect

function ismatch(pat, seq, allowed_mismatches)
	check = isequal.(pat, seq) .== false
	return count(check) <= allowed_mismatches
end

function find_all_matches(pat, seq, allowed_mismatches)

	padding = fill('x', length(pat) - 1) # In this case, padding allows xxA to match at the front, and Gxx to match at the back.
	seq_padded = [padding; seq; padding]
	ov = OffsetVector(seq_padded, (1:length(seq_padded)) .- length(padding))

	# Slide ismatch function accross all valid indexes.
	matches = map(firstindex(ov):lastindex(seq)) do i
		sub = i:i+length(pat) - 1
		if ismatch(pat, ov[sub], allowed_mismatches)
			return sub
		else
			return nothing
		end
	end

end
```

Replicating test.
```bash
julia> find_all_matches(pat, seq, 0)
9-element Vector{Union{Nothing, UnitRange{Int64}}}:
 nothing
 nothing
 1:3
 nothing
 ⋮
 5:7
 nothing
 nothing

julia> find_all_matches(pat, seq, 1)
9-element Vector{Union{Nothing, UnitRange{Int64}}}:
 nothing
 nothing
 1:3
 nothing
 ⋮
 5:7
 nothing
 nothing

julia> find_all_matches(pat, seq, 2)
9-element Vector{Union{Nothing, UnitRange{Int64}}}:
 nothing
 nothing
 1:3
 nothing
 ⋮
 5:7
 nothing
 nothing
```

Matching first and last symbol.
```bash
julia> find_all_matches(collect("xxA"), seq, 2)
9-element Vector{Union{Nothing, UnitRange{Int64}}}:
 -1:1
 0:2
 nothing
 nothing
 ⋮
 nothing
 nothing
 7:9

julia> find_all_matches(collect("Gxx"), seq, 2)
9-element Vector{Union{Nothing, UnitRange{Int64}}}:
 -1:1
 nothing
 nothing
 nothing
 ⋮
 nothing
 6:8
 7:9
```

Does it make sense to add padding and return adjusted ranges?

**Comment by CiaranOMara** [comment]
6 hours of hours later, I reckon the padding does make sense - I'm thinking cutadapt and Trimmomatic type stuff. But the length adjusted range was unexpected to me.

**Comment by CiaranOMara** [comment]
Is the current is approx search implemented as a distance-based search with an allowed number of edits?

**Comment by CiaranOMara** [comment]
I now see that that the name of the approx search method is inspired by the title of Gene Myers' paper, and returns match within a specified Levenshtein distance. 

I'll hide my last few comments as off-topic.

**Comment by BenJWard** [comment]
@jakobnissen Thanks for reviewing, I'll go though the comments and make any changes ASAP. I have been thinking about doing something about the code dup and how to do it, but I'm getting to the point where I really just want to have v3 and get the new Kmers.jl out.

Especially as for advanced searching, I wonder if we should be looking into providing and pushing familiarity with packages that build index structures (SuffixArrays, SuffixTrees etc.) and other data structures that are often the starting points HSP hit seeds etc. of more involved mappers anyway.

**Comment by CiaranOMara** [comment]
@BenJWard, what are the ways of deduplication that you've been thinking about, and could they affect the API?

As I've said before, I think subtyping the predicates from `AbstractQuery` or `AbstractSearchQuery` will help capture much of the basic duplication now. And the subtyping could be useful for further generalisation at some other time. It's not clear to me whether this idea is out, in now, or in later.

Are there tasks/units of work that are better to include early for v3 that you can sub out? I'd be keen to engage with repository project and discussion boards.

**Comment by BenJWard** [comment]
> @BenJWard Okay a few comments
> 
> First, I think this is great. It's much better than the old API, and we should use this. I have a few comments - most of them do not pertain to the API but are internal, so I can add that later as PRs if I can figure out how to. In no particular order
> 
>     * Any reason `k` must be specified when creating an `ApproxSearchQuery`? The problem with that is that I can easily imagine a user wanting to search with the same query with first 0, then 1, then 2 etc. mismatches until they find a hit. With this API, they need to create N number of queries. If the motivation is to keep the find API more similar to Base (which I support!), then perhaps add a function to create a query from another query with a new K? This can re-use all the internal data except K and so be almost instant. The alternative is to specify `k` in the find* functions.
> 
>     * You are right there are some code duplication. I think it should be possible to collapse all 4 `quicksearch` methods into one. But as you say, that is not critical and can be done later.
> 
>     * The `quicksearch` algorithm used for exact match seem to work best when the alphabet is larger, because it allows it to skip more symbols ahead when a match is rejected. That is unfortunate, since RNA and DNA alphabets are small. For example, I just benchmarked finding a short sequence after 70,000 symbols, where it is 7.2 times faster to search for amino acids than DNA. In fact, for exact DNA/RNA queries, I think it's faster to pack the first N symbols into an integer (just like a kmer, just simpler) and check that before continuing the search. I can look into that later, the speed is not too important now!
> 
>     * It would be nice to have an approximate search query that did not use compat bits but required an exact match. I think we should be able to do it just like you said, by generalizing over `isequal` vs `iscompatible` without too much hassle.
> 
>     * A small nitpick: Right now, the `ExactSearchQuery` includes the compatible query (which makes it not necessarily exact), whereas the `MarchSearchQuery` actually needs it to be exact. Doesn't it make more sense to switch the names of those two, such that the exact query means exact, and match is ambiguous to mean either exact match or compatible match?

@jakobnissen 

To your first point, `k` has been separated from the ApproxSearchQuery.
To your last point, I've swapped the abstract and concrete type names.

**Comment by BenJWard** [comment]
The Exact and Match SearchQuery types have been merged into one: ExactSearchQuery, that is parametric over the sequence type of the pattern, as well as a comparator function, so now you can do exact searches using either `isequal` or `iscompatible`. Making the concept much simpler, and cutting code in half.

**Comment by SabrinaJaye** [comment]
@jakobnissen Ok, and now the ApproxSearchQuery can be given `isequal` or `iscompatible` just as `ExactSearchQuery` can.

**Comment by SabrinaJaye** [comment]
@jakobnissen @CiaranOMara I now believe that this is good enough to merge and begin getting ready to release and merge v3.

File: src/search/pwm.jl [hunk 3]
```diff
@@ -315,3 +354,42 @@ function show_nuc_matrix(io::IO, m::Union{PFM{S},PWM{S}}) where S<:Union{DNA,RNA
...
 end
+
+struct PWMSearchQuery{S,T,R<:Real}
+    pwm::PWM{S,T}
+    threshold::R
+end
+
+function PWMSearchQuery(pwm::PWM{S,T}, threshold::R) where {S,T,R<:Real}
+    return PWMSearchQuery{S,T,R}(pwm, threshold)
+end
```
Comment (jakobnissen) [inline src/search/pwm.jl:365]: This function is redundant - it is automatically generated by the struct definition. Can be deleted

File: src/search/ExactSearchQuery.jl [hunk 4]
```diff
@@ -0,0 +1,200 @@
+"""
+Query type for exact sequence search.
+"""
+struct ExactSearchQuery{F<:Function,S<:BioSequence}
```
Comment (jakobnissen) [inline src/search/ExactSearchQuery.jl:4]: Would be nice with a simple docstring for this type, e.g. how to construct it, maybe even doctests.

File: src/search/ApproxSearchQuery.jl [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+#
+# Approximate sequence search tools.
+#
+# This file is a part of BioJulia.
+# License is MIT: https://github.com/BioJulia/BioSequences.jl/blob/master/LICENSE.md
+
+"""
+Query type for approximate sequence search.
+"""
+struct ApproximateSearchQuery{F<:Function,S<:BioSequence}
```
Comment (jakobnissen) [inline src/search/ApproxSearchQuery.jl:12]: Same as exact query: Would be nice with a simple docstring for this type, e.g. how to construct it, maybe even doctests.

**Comment by jakobnissen** [comment]
There are still some old bindings that are exported, but not used, like `approxrsearch`. These exports should be removed.

File: src/search/ExactSearchQuery.jl [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+function Base.findprev(query::ExactSearchQuery, seq::BioSequence, start::Integer)
+    i = quickrsearch(query, seq, start, 1)
+    if i == 0
+        return nothing
+    else
+        return i:i+length(query.seq)-1
+    end
+end
+
+Base.findlast(query::ExactSearchQuery, seq::BioSequence) = findprev(query, seq, lastindex(seq))
```
Comment (CiaranOMara) [inline src/search/ExactSearchQuery.jl:192]: I think all the `findlast` could become
```julia
Base.findlast(args...) = Base.findprev(args..., lastindex(seq))
```

Similar with `findfirst`.
```julia
Base.findfirst(args...) = Base.findnext(args..., firstindex(seq))
```

**Review by CiaranOMara — COMMENTED** [review]
I did not find any code issues. My comments are documentation suggestions. You've put together a much neater interface.

Concerning doctests, they could be added to the unit tests(https://github.com/BioJulia/GenomicFeatures.jl/blob/06f73422ffe134422e503fb0bc37ee75ca529d63/test/runtests.jl#L403-L405).

File: docs/src/sequence_search.md [hunk 7]
```diff
@@ -44,93 +63,88 @@ julia> occursin(query, seq)
...
 ```
 
-These search functions take ambiguous symbols into account.
-That is, if two symbols are compatible (e.g. `DNA_A` and `DNA_N`),
-they match when searching an occurrence.
-In the following example, 'N' is a wild card that matches any symbols.
+By default `ExactSearchQuery` uses `isequal` as the function to compare symbols.
+However in biology, sometimes we want to be a little bit flexible, and perhaps
+find subsequences where the symbols are _compatible_ with each other. To do this,
+we can specify the `iscompatible` function in the constructor of `ExactSearchQuery`.
```
[the remark this answers is not in the document]
Comment (SabrinaJaye) [reply docs/src/sequence_search.md:69]: > There are still some old bindings that are exported, but not used, like `approxrsearch`. These exports should be removed. Here is a list of exported but not defined symbols: [:Ambiguous, :Certain, :Gap, :Match, :Mismatch, :Site, :approxrsearch, :approxrsearchindex, :approxsearch, :approxsearchindex, :count_pairwise, :each, :eachcanonical]

These have been removed

**Pull request merged**